compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLPatterns.cpp

```diff
@@ -566,6 +574,84 @@ class ApplyPDLPatternsPass
     }
   }
 
+private:
+  /// Loaded PDL patterns
+  FrozenRewritePatternSet patterns;
+};
+} // namespace
+
+// Populate patterns from PDLL files.
+static LogicalResult
+populatePDLLModuleFromFileName(MLIRContext *context,
+                               RewritePatternSet &patterns,
+                               llvm::StringRef pdllModuleFileName) {
+  std::string errorMessage;
+  auto memoryBuffer = mlir::openInputFile(pdllModuleFileName, &errorMessage);
+  if (!memoryBuffer) {
+    return emitError(FileLineColLoc::get(
+               StringAttr::get(context, pdllModuleFileName), 0, 0))
+           << "failed to open pattern module file: " << errorMessage;
+  }
+  // Tell sourceMgr about this buffer, the parser will pick it up.
+  llvm::SourceMgr sourceMgr;
+  sourceMgr.AddNewSourceBuffer(std::move(memoryBuffer), llvm::SMLoc());
+
+  // parse file and build pdll ast module
+  pdll::ods::Context pdllOdsContext;
+  pdll::ast::Context pdllAstContext(pdllOdsContext);
+  FailureOr<pdll::ast::Module *> pdllModule =
+      pdll::parsePDLLAST(pdllAstContext, sourceMgr);
+  if (failed(pdllModule)) {
+    return failure();
+  }
+
+  PDLPatternModule pdlModule =
+      pdll::codegenPDLLToMLIR(context, pdllAstContext, sourceMgr, **pdllModule);
+  pdlModule.registerRewriteFunction("rewriteAsFlowDispatch",
+                                    rewriteAsFlowDispatch);
+  pdlModule.registerConstraintFunction("checkTensorElementType",
+                                       checkTensorElementType);
+  patterns.insert(std::move(pdlModule));
+  return success();
+}
+
+namespace {
+
+class ApplyPDLLPatternsPass
```

Comment:
We can move this pass to a separate file?

Reply:
In this case, we will need to move all the common functions to a utility header.